### DISCUSSION TITLE
Cross ref autocompletions don't work in VS Code in some projects

### DISCUSSION THREAD

**Discussion by pmc4** [quarto-dev/quarto-cli#6844] [Q&A]
### Description

Hi everyone! 

I'm using Quarto for creating my PhD notes. I created a project and set a template and some latex code for it. I have several .qmd files, one for each chapter, that are included in the _quarto.yml file. Inside those I also use the include command, e.g. `{{< include frontmatter/preface.qmd >}}`.

My problem is that when typing `@` for referencing, I don't get any autocompletion suggestion. I've checked that the references are properly set with `#eq-` or `#tbl-` and so on. I've also discovered that if I create a completely new Quarto document (not inside the same project) and save it, the autocompletion then works.

Why is therefore not working in my PhD notes project? Could it be because the file structure is more complex? I'm using quarto v1.3.450 and the last vs code extension version.

If anyone has any idea why this is happening, I would appreciate it.

P.s.: if you need any file or something, tell me. I haven't sent them because I wanted to keep the question short.

**Comment by mcanouil** [comment]
Without some Git repository showing the issue for us to investigate, I am afraid, it's unlikely we'll luckily guess what's the issue.

So, could you share a small reproducible example that exhibit the issue? Since you are saying that autocompletion works for single Quarto document, it might have to do with your project structure, so if you can mimic your Ph.D. Quarto project structure while keeping it as small as possible it would help us help you.

**Reply by pmc4** [reply]
I have reduced the project to the bare minimum. From here, expanding with new parts and chapters is easy. I include only one part with one chapter. Inside that you can see an explanation of the project structure and also an example of a table and an equation. The cross reference of such work, but the autocompletion inside .vscode doesn't.

This is the project as a compressed file:
[quarto-book-project.tar.gz](https://github.com/quarto-dev/quarto-cli/files/12618713/quarto-book-project.tar.gz)

**Reply by mcanouil** [reply]
Could you share this as a Git repository as I requested? (it's easy to clone if we need to, and it's also easy to look at the files directly, which is not at all possible with an archive amongst other considerations)

**Reply by pmc4** [reply]
Oh, sorry. I created one for the minimal version:
https://github.com/pmc4/phd-notes

**Reply by pmc4** [reply]
Hi! Sorry for bothering you, could you find any hint about the problem?

**Reply by mcanouil** [reply]
In VSCode, you get the auto-completion if you place `bibliography` at the top level and not under the format.

@jjallaire is it an expected behaviour?

https://github.com/quarto-dev/quarto-cli/assets/8896044/8005d7c8-c2c6-4808-af35-fb2227bc033c

**Comment by jjallaire** [comment]
Yes this is indeed expected based on the way we parse the bibliography from YAML (not ideal but that is the way it currently works)

**Reply by mcanouil** [reply]
It actually makes more sense to me to set the bibliography at the top level rather than under the format which imply you might want to set different bibliography files for different format of the same document which is very unlikely to ever happen.
Maybe, Quarto could recommends to set `bibliography` at the top level via yaml validation?

CC @cscheid

**Reply by pmc4** [reply]
Putting the bibliography at the top level as @mcanouil showed in the video worked, thanks! Still, with cross references (equations, tables and so on) I don't get the autocompletion feature. Even if they are defined on the same .qmd document.

**Comment by jjallaire** [comment]
I can repro the issue with cross references locally (thanks for the git repo!). Investigating now....

**Comment by jjallaire** [comment]
Okay, there are two issues here (neither of which we will have a short term fix for). Hopefully you can still work around them though.

1) If there is no `html` format in a book we don't create the crossref index used by completion

2) Crossrefs within `include` files are not indexed for completion.

**Reply by pmc4** [reply]
I don't mind going to the definition of the equation and read the label. Since these are some notes and not the thesis, each chapter will be almost self-contained. Regarding your points:

1. If I add the html format, but only render it in pdf, would I be able to see the cross refs inside the same .qmd file?
2. It would be nice to have that feature, but it's alright.

**Reply by jjallaire** [reply]
Sadly you do need to render to HTML to get the _global_ crossref index populated. However, you should get crossrefs that are local to a file even if you never render it to HTML

**Reply by pmc4** [reply]
I see. I'll have to work with that in the meantime. Thanks for the help!
